ledger_device_sdk/src/io_new.rs

```diff
@@ -0,0 +1,650 @@
+            }
+            (len, _) => {
+                if rx_len != len as usize + 5 {
+                    return DecodedEventType::ApduError(BadLen);
+                }
+                DecodedEventType::new_apdu(header, 1 + 5, len as usize)
+            }
+        }
+    }
+}
+
+/// High-level decoded event (no side effects).
+pub enum DecodedEventType {
+    Apdu {
+        header: ApduHeader,
+        offset: usize,
+        length: usize,
+    },
+    ApduError(ApduError),
+    #[cfg(not(any(target_os = "stax", target_os = "flex", target_os = "apex_p")))]
+    Button(ButtonEvent),
+    #[cfg(any(target_os = "stax", target_os = "flex", target_os = "apex_p"))]
+    Touch,
+    Ticker,
+    // Events for which no additional handling is required after decoding it
+    Ignored,
+}
+
+impl DecodedEventType {
+    fn new_apdu(header: ApduHeader, offset: usize, length: usize) -> Self {
+        Self::Apdu {
+            header,
+            offset,
+            length,
+        }
+    }
+}
+
+pub struct Command<'a, const N: usize = DEFAULT_BUF_SIZE> {
+    comm: &'a mut Comm<N>,
+    header: ApduHeader,
+    offset: usize,
+    length: usize,
+}
+
+impl<'a, const N: usize> Command<'a, N> {
+    pub fn new(comm: &'a mut Comm<N>, header: ApduHeader, offset: usize, length: usize) -> Self {
+        Self {
+            comm,
+            header,
+            offset,
+            length,
+        }
+    }
+
+    pub fn decode<T>(&self) -> Result<T, Reply>
+    where
+        T: TryFrom<ApduHeader>,
+        Reply: From<<T as TryFrom<ApduHeader>>::Error>,
+    {
+        T::try_from(self.header).map_err(Reply::from)
+    }
+
+    pub fn get_data(&self) -> &[u8] {
+        &self.comm.buf[self.offset..self.offset + self.length]
+    }
+
+    pub fn into_tx(self) -> Tx<'a, N> {
+        Tx {
+            comm: self.comm,
+            len: 0,
+        }
+    }
+
+    pub fn into_comm(self) -> &'a mut Comm<N> {
+        self.comm
+    }
+
+    pub fn reply<T: Into<Reply>>(self, data: &[u8], reply: T) -> Result<(), CommError> {
+        self.into_tx().extend(data)?.send(reply)?;
+        Ok(())
+    }
+}
+
+/// Immutable read view.
+pub struct Rx<'a, const N: usize = DEFAULT_BUF_SIZE> {
+    comm: &'a mut Comm<N>,
+    len: usize,
+}
+
+impl<'a, const N: usize> core::ops::Deref for Rx<'a, N> {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl<'a, const N: usize> Rx<'a, N> {
+    pub fn as_slice(&self) -> &[u8] {
+        &self.comm.buf[..self.len]
+    }
+    pub fn len(&self) -> usize {
+        self.len
+    }
+    pub fn consume(self) -> () {}
+
+    /// Decode into a higher-level event. No replies are sent, but UX-related and other OS interactions are dealt with.
+    pub fn decode_event(self) -> DecodedEvent<N> {
+        DecodedEvent::new(self.comm, self.len)
+    }
+}
+
+/// Mutable write view for building a send.
+pub struct Tx<'a, const N: usize = DEFAULT_BUF_SIZE> {
```

Comment:
Same comment as `Rx`

Reply:
As this is an object that apps would use, renamed to `CommandResponse` in 51e82242cc673d29f3b7334e647ae36ddcf558dc (and renamed `begin_tx` to `begin_response`).